frontend/src/metabase/browse/containers/BrowseModels.tsx

```diff
@@ -0,0 +1,278 @@
+    b.collection?.name?.toLowerCase() || fallbackSortValue;
+
+  if (collectionNameA < collectionNameB) {
+    return -1;
+  }
+  if (collectionNameA > collectionNameB) {
+    return 1;
+  }
+
+  // If the two models' parent collections have the same name, sort on the id of the collection
+  const collectionIdA = a.collection?.id ?? fallbackSortValue;
+  const collectionIdB = b.collection?.id ?? fallbackSortValue;
+
+  if (collectionIdA < collectionIdB) {
+    return -1;
+  }
+  if (collectionIdA > collectionIdB) {
+    return 1;
+  }
+
+  const nameA = a.name.toLowerCase() || fallbackSortValue;
+  const nameB = b.name.toLowerCase() || fallbackSortValue;
+
+  // If the two collection ids are the same, sort on the names of the models
+  if (nameA < nameB) {
+    return -1;
+  }
+  if (nameA > nameB) {
+    return 1;
+  }
+
+  // If the two models have the same name, sort on id
+  const idA = a.id ?? fallbackSortValue;
+  const idB = b.id ?? fallbackSortValue;
+
+  if (idA < idB) {
+    return -1;
+  }
+  if (idA > idB) {
+    return 1;
+  }
+
+  return 0;
+};
+
+const CollectionHeader = ({
+  collection,
+  style,
+  id,
+}: {
+  collection?: Pick<Collection, "id" | "name"> | null;
+  style?: React.CSSProperties;
+  id: string;
+}) => {
+  const MaybeLink = ({ children }: { children: React.ReactNode }) =>
+    collection ? (
+      <Link to={Urls.collection(collection)}>{children}</Link>
+    ) : (
+      <>{children}</>
+    );
+  return (
+    <CollectionHeaderContainer id={id} role="heading" style={style}>
+      <MaybeLink>
+        <Icon
+          name="folder"
+          color={"text-dark"}
+          size={16}
+          style={{ marginRight: "0.33rem" }}
+        />
+        <h4>{collection?.name || "Untitled collection"}</h4>
+      </MaybeLink>
+    </CollectionHeaderContainer>
+  );
+};
+
+type Cell = React.ReactElement | null;
+
+const makeCells = (models: SearchResult[]): Cell[] => {
+  const cells: Cell[] = [];
+  for (let i = 0; i < models.length; i++) {
+    const model = models[i];
+
+    const collectionIdChanged =
+      models[i - 1]?.collection?.id !== model.collection?.id;
+
+    const firstModelInItsCollection =
+      i === 0 || collectionIdChanged || model.collection?.id === undefined;
+
+    /** This id is used by aria-labelledby */
+    const collectionHtmlId = model?.collection?.id
+      ? `collection-${model.collection?.id}`
+      : `item-${cells.length}`;
+
+    // Before the first model in a given collection,
+    // add an item that represents the header of the collection
+    if (firstModelInItsCollection) {
+      const header = (
+        <CollectionHeader
+          collection={model.collection}
+          key={collectionHtmlId}
+          id={collectionHtmlId}
+        />
+      );
+      cells.push(header);
+    }
+
+    cells.push(
+      <ModelCell
+        collectionHtmlId={collectionHtmlId}
+        key={`model-${model.id}`}
+        model={model}
+      />,
+    );
+  }
+  return cells;
+};
```

Comment:
We could also create this list of `<ModelCells>` in a loop inside some jsx, but this approach makes it much easier to drop in virtualization in [the next PR](https://github.com/metabase/metabase/pull/37818)